### PR TITLE
feat: [CDS-70209]: Validating multiple helm chart manifests

### PR DESCRIPTION
**Pull request by GoelPratyush99**
## Describe your changes
Added support to store multiple Helm Chart (only) Manifests in ServiceV2.  The user would need to enable the FF: `CDS_HELM_MULTIPLE_MANIFEST_SUPPORT_NG` and have `manifestConfigurations` field in service yaml to support this feature. We do not support this feature for Service V1.

Local Testing in progress will update Screen Shots
Screenshot 2023-07-17 at 3 59 44 PM

Screenshot 2023-07-17 at 3 59 51 PM

Screenshot 2023-07-17 at 4 00 10 PM

When FF is disabled but service yaml contains manifest configurations.
Screenshot 2023-07-17 at 4 25 57 PM

## Checklist
- [x] I've documented the changes in the PR description.
- [x] I've tested this change either in PR or local environment.
- [x] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/50559)

### CODE REVIEW

File: 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceApiUtils.java [hunk 1]
```diff
@@ -89,32 +92,10 @@ public ServiceResponse mapToAccessListResponse(ServiceEntity serviceEntity) {
...
-    if (ngServiceV2InfoConfig.getServiceDefinition() != null) {
-      serviceEntity.setType(ngServiceV2InfoConfig.getServiceDefinition().getType());
-    }
-    Set<ConstraintViolation<ServiceEntity>> violations = validator.validate(serviceEntity);
-    if (!violations.isEmpty()) {
-      throw new JerseyViolationException(violations, null);
-    }
-
-    return serviceEntity;
+    serviceSchemaHelper.throwExceptionForInvalidServiceEntity(serviceEntity, account);
```
Comment (abhinavsingh1196) [inline 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceApiUtils.java:97]: serviceEntity already contains `account`, no need to pass it separately.
Also, this call `serviceSchemaHelper.throwExceptionForInvalidServiceEntity` should go into the below method  ` validateAndUpdateServiceEntity` since schema helper is also performing validation?

File: 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceApiUtils.java [hunk 2]
```diff
@@ -41,10 +42,12 @@
 @Singleton
 public class ServiceResourceApiUtils {
   private final Validator validator;
+  private ServiceEntityYamlSchemaHelper serviceSchemaHelper;
```
Comment (abhinavsingh1196) [inline 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceApiUtils.java:45]: nit: can be final?

File: 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceV2.java [hunk 3]
```diff
@@ -276,7 +277,9 @@ public class ServiceResourceV2 {
         ResourceScope.of(accountId, serviceRequestDTO.getOrgIdentifier(), serviceRequestDTO.getProjectIdentifier()),
         Resource.of(NGResourceType.SERVICE, null), SERVICE_CREATE_PERMISSION);
     serviceSchemaHelper.validateSchema(accountId, serviceRequestDTO.getYaml());
-    ServiceEntity serviceEntity = ServiceElementMapper.toServiceEntity(accountId, serviceRequestDTO);
+    ServiceEntity serviceEntity = ServiceElementMapper.mapServiceResponseToServiceEntity(serviceRequestDTO, accountId);
+    serviceSchemaHelper.throwExceptionForInvalidServiceEntity(serviceEntity, accountId);
+    serviceEntity = ServiceElementMapper.validateAndUpdateServiceEntity(serviceEntity);
```
Comment (abhinavsingh1196) [inline 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceV2.java:282]: this snippet could be a method of its own with an appropriate name to avoid repetition?

ServiceEntity serviceEntity = ServiceElementMapper.mapServiceResponseToServiceEntity(serviceRequestDTO, accountId);
serviceSchemaHelper.throwExceptionForInvalidServiceEntity(serviceEntity, accountId);
return ServiceElementMapper.validateAndUpdateServiceEntity(serviceEntity);

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java [hunk 4]
```diff
@@ -194,15 +196,23 @@ public ChildrenExecutableResponse obtainChildren(
...
-          "Service Name: " + servicePartResponse.getNgServiceConfig().getNgServiceV2InfoConfig().getName()
-              + " , Identifier: "
-              + servicePartResponse.getNgServiceConfig().getNgServiceV2InfoConfig().getIdentifier());
+          "Service Name: " + ngServiceV2InfoConfig.getName()
+              + " , Identifier: " + ngServiceV2InfoConfig.getIdentifier());
+
+      performFiltering(
+          stepParameters, ambiance, EnvironmentInfraFilterHelper.getNGTags(ngServiceV2InfoConfig.getTags()));
+
+      if (ambiance != null && ngServiceV2InfoConfig.getServiceDefinition() != null
```
Comment (abhinavsingh1196) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java:207]: nit: can we rewrite this block to something like 

Optional<ManifestConfigurations> manifestConfigurationsOpt = getManifestConfigurations(ambiance, ngServiceV2InfoConfig)
if (manifestConfigurationsOpt.isPresent()) {
  resolve(ambiance, manifestConfigurationsOpt.get().getPrimaryManifestRef())
}


Optional<ManifestConfigurations> getManifestConfigurations(ambiance, ngServiceV2InfoConfig) {
  if (ambiance == null || ngServiceV2InfoConfig.getServiceDefinition() == null || ngServiceV2InfoConfig.getServiceDefinition().getServiceSpec() == null) {
    return Optional.empty()
  }

  return ServiceFilterHelper.getManifestConfigurations(ngServiceV2InfoConfig.getServiceDefinition().getServiceSpec())
}

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/services/impl/ServiceEntityYamlSchemaHelper.java [hunk 5]
```diff
@@ -63,4 +68,13 @@ && isNotEmpty(yaml)) {
       }
     }
   }
+
+  public void throwExceptionForInvalidServiceEntity(ServiceEntity serviceEntity, String accountId) {
```
Comment (abhinavsingh1196) [inline 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/services/impl/ServiceEntityYamlSchemaHelper.java:72]: nit: can be renamed to something like `throwExceptionIfPrimaryManifestNotAllowed`

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/NGServiceEntityMapper.java [hunk 6]
```diff
@@ -97,7 +107,75 @@ private static void validateManifests(ServiceDefinition serviceDefinition) {
                                                   .map(ManifestConfig::getSpec)
                                                   .filter(Objects::nonNull)
                                                   .collect(Collectors.toList());
-      SvcEnvV2ManifestValidator.validateManifestList(serviceDefinition.getType(), manifestList);
+      ManifestConfigurations manifestConfigurations =
+          ServiceFilterHelper.getManifestConfigurations(serviceDefinition.getServiceSpec());
+      if (manifestConfigurations == null || ParameterField.isNull(manifestConfigurations.getPrimaryManifestRef())) {
```
Comment (abhinavsingh1196) [inline 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/NGServiceEntityMapper.java:112]: this can go inside the below function `validateManifestList`
  ManifestConfigurations manifestConfigurations =
          ServiceFilterHelper.getManifestConfigurations(serviceDefinition.getServiceSpec());
      if (manifestConfigurations == null || ParameterField.isNull(manifestConfigurations.getPrimaryManifestRef())) {
       SvcEnvV2ManifestValidator.validateManifestList(serviceDefinition.getType(), manifestList);
       return;
}

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/NGServiceEntityMapper.java [hunk 7]
```diff
@@ -97,7 +107,75 @@ private static void validateManifests(ServiceDefinition serviceDefinition) {
...
+    try {
+      final NGServiceConfig config = YamlPipelineUtils.read(serviceEntity.getYaml(), NGServiceConfig.class);
+      if (config == null || config.getNgServiceV2InfoConfig() == null
+          || config.getNgServiceV2InfoConfig().getServiceDefinition() == null
+          || config.getNgServiceV2InfoConfig().getServiceDefinition().getServiceSpec() == null) {
+        return false;
+      }
+      ManifestConfigurations manifestConfigurations = ServiceFilterHelper.getManifestConfigurations(
+          config.getNgServiceV2InfoConfig().getServiceDefinition().getServiceSpec());
+      return manifestConfigurations != null && !ParameterField.isNull(manifestConfigurations.getPrimaryManifestRef());
```
Comment (abhinavsingh1196) [inline 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/NGServiceEntityMapper.java:176]: this logic looks pretty similar to 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java
can we make a common method for this?

**Review by abhinavsingh1196 — CHANGES_REQUESTED** [review]

File: 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceApiUtils.java [hunk 8]
```diff
@@ -188,4 +169,36 @@ public List<ServiceResponse> filterByPermissionAndId(
     }
     return filteredAccessControlDtoList;
   }
+
+  private static ServiceEntity mapServiceResponseToServiceEntity(
```
Comment (yogesh-chauhan) [inline 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceApiUtils.java:173]: nit: need not be static

File: 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceApiUtils.java (hunk 1 above)
Comment (yogesh-chauhan) [reply 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceApiUtils.java:97]: +1

File: 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceV2.java [hunk 9]
```diff
@@ -230,6 +230,7 @@ public class ServiceResourceV2 {
         serviceEntityService.get(accountId, orgIdentifier, projectIdentifier, serviceIdentifier, deleted);
     if (serviceEntity.isPresent()) {
       if (EmptyPredicate.isEmpty(serviceEntity.get().getYaml())) {
+        serviceSchemaHelper.throwExceptionForInvalidServiceEntity(serviceEntity.get(), accountId);
```
Comment (yogesh-chauhan) [inline 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceV2.java:233]: no need of this check here.

Also this method is validating manifests but its name is very very generic, which is confusing

File: 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceV2.java [hunk 10]
```diff
@@ -311,7 +314,12 @@ public class ServiceResourceV2 {
         serviceRequestDTO -> serviceSchemaHelper.validateSchema(accountId, serviceRequestDTO.getYaml()));
     List<ServiceEntity> serviceEntities =
         serviceRequestDTOs.stream()
-            .map(serviceRequestDTO -> ServiceElementMapper.toServiceEntity(accountId, serviceRequestDTO))
+            .map(serviceRequestDTO -> {
+              ServiceEntity serviceEntity =
+                  ServiceElementMapper.mapServiceResponseToServiceEntity(serviceRequestDTO, accountId);
+              serviceSchemaHelper.throwExceptionForInvalidServiceEntity(serviceEntity, accountId);
+              return ServiceElementMapper.validateAndUpdateServiceEntity(serviceEntity);
```
Comment (yogesh-chauhan) [inline 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceV2.java:321]: looks like repeated code, maybe extract out a method ?

Comment (yogesh-chauhan) [reply 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceV2.java:321]: Also we already have "serviceSchemaHelper.validateSchema", can we put the manifest logic there ?

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepOverrideHelper.java [hunk 11]
```diff
@@ -675,4 +681,21 @@ private boolean isNoEnvGlobalConnectionStringsOverridePresent(NGEnvironmentConfi
...
+      return emptyIfNull(svcManifests);
+    }
+    String primaryManifestId =
+        ParameterFieldHelper.getParameterFieldValue(manifestConfigurations.getPrimaryManifestRef());
+    if (isEmpty(primaryManifestId)) {
+      return emptyIfNull(svcManifests);
+    }
+    return svcManifests.stream()
+        .filter(manifest
+            -> SERVICE_OVERRIDE_SUPPORTED_MANIFEST_TYPES.contains(manifest.getManifest().getType().getDisplayName())
```
Comment (yogesh-chauhan) [inline 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepOverrideHelper.java:697]: why are we filtering on SERVICE_OVERRIDE_SUPPORTED_MANIFEST_TYPES?

**Review by yogesh-chauhan — CHANGES_REQUESTED** [review]

File: 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceApiUtils.java (hunk 1 above)
Comment (GoelPratyush99) [reply 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceApiUtils.java]: removed it

File: 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceApiUtils.java (hunk 2 above)
Comment (GoelPratyush99) [reply 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceApiUtils.java]: done

File: 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceV2.java (hunk 3 above)
Comment (GoelPratyush99) [reply 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceV2.java]: Moved it to ServiceEntityYamlSchemaHelper class

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java (hunk 4 above)
Comment (GoelPratyush99) [reply 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepV3.java]: done

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/services/impl/ServiceEntityYamlSchemaHelper.java (hunk 5 above)
Comment (GoelPratyush99) [reply 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/services/impl/ServiceEntityYamlSchemaHelper.java]: done

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/NGServiceEntityMapper.java (hunk 6 above)
Comment (GoelPratyush99) [reply 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/NGServiceEntityMapper.java]: done

File: 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceApiUtils.java (hunk 8 above)
Comment (GoelPratyush99) [reply 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceApiUtils.java]: removed it

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/NGServiceEntityMapper.java (hunk 7 above)
Comment (GoelPratyush99) [reply 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/NGServiceEntityMapper.java]: done

File: 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceV2.java (hunk 9 above)
Comment (GoelPratyush99) [reply 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceV2.java]: renamed it to throwExceptionIfPrimaryManifestNotAllowed

File: 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceV2.java (hunk 10 above)
Comment (GoelPratyush99) [reply 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceV2.java]: We need to return a service entity so I have created a separate method.

**Comment by GoelPratyush99** [comment]
trigger uts

File: 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepOverrideHelper.java (hunk 11 above)
Comment (GoelPratyush99) [reply 125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/service/steps/ServiceStepOverrideHelper.java:692]: We need not lose values the service override manifest while filtering the primary manifest. After refactoring it should make more sense.

**Comment by GoelPratyush99** [comment]
trigger uts

**Comment by GoelPratyush99** [comment]
trigger ut9_JDK11

**Comment by GoelPratyush99** [comment]
trigger uts

File: 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/AbstractServicesApiImpl.java [hunk 12]
```diff
@@ -102,6 +102,7 @@ public Response getServiceEntity(String org, String project, String service, Str
     }
     ServiceEntity optionalServiceEntity = serviceEntity.get();
     if (EmptyPredicate.isEmpty(serviceEntity.get().getYaml())) {
+      serviceSchemaHelper.throwExceptionIfPrimaryManifestNotAllowed(optionalServiceEntity);
```
Comment (yogesh-chauhan) [inline 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/AbstractServicesApiImpl.java:105]: we do not need this here

**Review by yogesh-chauhan — CHANGES_REQUESTED** [review]

**Comment by GoelPratyush99** [comment]
trigger uts

File: 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/AbstractServicesApiImpl.java (hunk 12 above)
Comment (GoelPratyush99) [reply 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/AbstractServicesApiImpl.java]: removed it

**Comment by GoelPratyush99** [comment]
trigger ut5, ut6, ut7, ut8, ut9

**Comment by GoelPratyush99** [comment]
trigger ut5, ut6, ut7, ut8, ut9

**Comment by GoelPratyush99** [comment]
trigger message-check

**Comment by GoelPratyush99** [comment]
trigger messagecheck

**Comment by GoelPratyush99** [comment]
trigger messagecheck

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/services/impl/ServiceEntityYamlSchemaHelper.java [hunk 13]
```diff
@@ -63,4 +70,20 @@ && isNotEmpty(yaml)) {
       }
     }
   }
+
+  public ServiceEntity getServiceEntityAndValidateItsFields(ServiceRequestDTO serviceRequestDTO, String accountId) {
```
Comment (abhinavsingh1196) [inline 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/services/impl/ServiceEntityYamlSchemaHelper.java:74]: nit: better name `validateAndGetServiceEntity` ?

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/ServiceElementMapper.java [hunk 14]
```diff
@@ -123,4 +105,30 @@ public ServiceBasicInfo toBasicInfo(ServiceEntity serviceEntity) {
...
+        .orgIdentifier(serviceRequestDTO.getOrgIdentifier())
+        .projectIdentifier(serviceRequestDTO.getProjectIdentifier())
+        .name(serviceRequestDTO.getName())
+        .description(serviceRequestDTO.getDescription())
+        .tags(convertToList(serviceRequestDTO.getTags()))
+        .yaml(serviceRequestDTO.getYaml())
+        .build();
+  }
+
+  public ServiceEntity validateAndUpdateServiceEntity(ServiceEntity serviceEntity) {
```
Comment (abhinavsingh1196) [inline 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/ServiceElementMapper.java:122]: is there any validation being done here? i don't see any. 
Also, this method is almost the same as 120-ng-manager/src/main/java/io/harness/ng/core/service/resources/ServiceResourceApiUtils#validateAndUpdateServiceEntity, except validator is being called there additionally. 

can we refactor out the common part and name it appropriately?

File: 127-cd-nextgen-entities/src/test/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImplTest.java [hunk 15]
```diff
@@ -1228,8 +1228,8 @@ public void testGetManifestIdentifiersListServiceV2KubernetesSpec() {
     String yaml = readFile(filename);
     ManifestsResponseDTO responseDTO = serviceEntityService.getManifestIdentifiers(yaml, SERVICE_ID);
     assertThat(responseDTO).isNotNull();
-    assertThat(responseDTO.getIdentifiers()).isNotNull().isNotEmpty().hasSize(2);
-    assertThat(responseDTO.getIdentifiers()).hasSameElementsAs(Arrays.asList("mani_i1", "mani_i2"));
```
Comment (abhinavsingh1196) [inline 127-cd-nextgen-entities/src/test/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImplTest.java:1232]: can there be additional tests added here? I don't like just editing tests to make them pass. if there are multiple scenarios, we should add tests for them if possible

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/NGServiceEntityMapper.java [hunk 16]
```diff
@@ -97,7 +110,84 @@ private static void validateManifests(ServiceDefinition serviceDefinition) {
...
+
+  private static void throwMultipleManifestsExceptionIfApplicable(
+      Map<String, String> manifestIdToManifestTypeMap, String deploymentType, Set<String> supported) {
+    if (manifestIdToManifestTypeMap.values().size() > 0) {
+      String manifestIdType = manifestIdToManifestTypeMap.entrySet()
+                                  .stream()
+                                  .map(entry -> String.format("%s : %s", entry.getKey(), entry.getValue()))
+                                  .collect(Collectors.joining(", "));
+      throw new InvalidRequestException(String.format(
+          "Manifests found [%s]. %s deployment with primaryManifestRef support only manifest of types: %s. Remove all other manifests",
```
Comment (abhinavsingh1196) [inline 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/NGServiceEntityMapper.java:160]: this message can be better? 
`Specifying multiple manifests using primaryManifestRef is only supported for the manifest types: [...]. Manifests found: [...]`

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/NGServiceEntityMapper.java [hunk 17]
```diff
@@ -97,7 +110,84 @@ private static void validateManifests(ServiceDefinition serviceDefinition) {
...
+
+  private static void validateMultipleManifests(
+      List<ManifestAttributes> manifestList, Set<String> supported, String deploymentType) {
+    final Map<String, String> manifestIdTypeMap =
+        manifestList.stream()
+            .filter(m -> supported.contains(m.getKind()) && !ManifestType.HelmChart.equals(m.getKind()))
+            .collect(Collectors.toMap(ManifestAttributes::getIdentifier, ManifestAttributes::getKind));
+
+    throwMultipleManifestsExceptionIfApplicable(
+        manifestIdTypeMap, deploymentType, Collections.singleton(ManifestType.HelmChart));
```
Comment (abhinavsingh1196) [inline 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/NGServiceEntityMapper.java:149]: should we have a static set elsewhere like `MULTIPLE_MANIFESTS_SUPPORTED_TYPES = [ManifestType.HelmChart]` ?

this filter can be generic then
` .filter(m -> supported.contains(m.getKind()) && !MULTIPLE_MANIFESTS_SUPPORTED_TYPES.equals(m.getKind()))`

and this method arg can also be replaced 
` throwMultipleManifestsExceptionIfApplicable(manifestIdTypeMap, deploymentType, MULTIPLE_MANIFESTS_SUPPORTED_TYPES);

**Review by abhinavsingh1196 — CHANGES_REQUESTED** [review]

**Comment by GoelPratyush99** [comment]
trigger uts

**Comment by GoelPratyush99** [comment]
trigger uts

**Comment by GoelPratyush99** [comment]
trigger uts

**Comment by GoelPratyush99** [comment]
trigger uts

**Comment by GoelPratyush99** [comment]
trigger uts

**Review by abosii-harness — DISMISSED** [review]

**Review by TarunUba1 — DISMISSED** [review]

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/NGServiceEntityMapper.java [hunk 18]
```diff
@@ -97,7 +107,80 @@ private static void validateManifests(ServiceDefinition serviceDefinition) {
                                                   .map(ManifestConfig::getSpec)
                                                   .filter(Objects::nonNull)
                                                   .collect(Collectors.toList());
-      SvcEnvV2ManifestValidator.validateManifestList(serviceDefinition.getType(), manifestList);
+      validateManifestList(serviceDefinition.getServiceSpec(), serviceDefinition.getType(), manifestList);
     }
   }
+
+  private static void validateManifestList(
```
Comment (tchaurasiya) [inline 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/NGServiceEntityMapper.java:114]: nit: private method need not to be static

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/ServiceFilterHelper.java [hunk 19]
```diff
@@ -306,11 +311,22 @@ public static void populateIdentifiersOfEachLevel(String accountIdentifier, Stri
     }
   }
 
-  public YamlField getManifestsNodeFromServiceDefinitionYaml(YamlField serviceDefinitionField) {
+  public YamlField getManifestsNodeFromServiceYaml(YamlField serviceYamlField) {
```
Comment (tchaurasiya) [inline 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/ServiceFilterHelper.java:314]: nit: if this method is not related to filters lets move it.

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/ServiceFilterHelper.java [hunk 20]
```diff
@@ -319,20 +335,30 @@ public YamlField getManifestsNodeFromServiceDefinitionYaml(YamlField serviceDefi
...
-      if (manifest != null && supportedManifestTypes.contains(manifest.getType())) {
+      if (manifest != null && ManifestType.HelmChart.equals(manifest.getType())) {
         String manifestIdentifier = manifest.getIdentifier();
         manifestIdentifierList.add(manifestIdentifier);
       }
     }
     return manifestIdentifierList;
   }
+
+  public ManifestConfigurations getManifestConfigurations(ServiceSpec spec) {
```
Comment (tchaurasiya) [inline 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/ServiceFilterHelper.java:354]: nit: if this method is not related to filters lets move it.

**Review by tchaurasiya — APPROVED** [review]

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/NGServiceEntityMapper.java [hunk 21]
```diff
@@ -97,7 +107,80 @@ private static void validateManifests(ServiceDefinition serviceDefinition) {
...
+        manifestList.stream()
+            .filter(m -> supported.contains(m.getKind()) && !MULTIPLE_SUPPORTED_MANIFEST_TYPES.contains(m.getKind()))
+            .collect(Collectors.toMap(ManifestAttributes::getIdentifier, ManifestAttributes::getKind));
+
+    throwMultipleManifestsExceptionIfApplicable(manifestIdTypeMap, deploymentType, MULTIPLE_SUPPORTED_MANIFEST_TYPES);
+  }
+
+  private static void throwMultipleManifestsExceptionIfApplicable(
+      Map<String, String> manifestIdToManifestTypeMap, String deploymentType, Set<String> supported) {
+    if (manifestIdToManifestTypeMap.values().size() > 0) {
```
Comment (tchaurasiya) [inline 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/NGServiceEntityMapper.java:150]: nit: can directly use EmptyPredicate.isNotEmpty()

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/ServiceFilterHelper.java [hunk 22]
```diff
@@ -319,20 +335,30 @@ public YamlField getManifestsNodeFromServiceDefinitionYaml(YamlField serviceDefi
...
+    if (spec instanceof KubernetesServiceSpec) {
```
Comment (tchaurasiya) [inline 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/mappers/ServiceFilterHelper.java:355]: Also can we add k8s/nativeHelm in method name, if we are only handling that

File: 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImpl.java [hunk 23]
```diff
@@ -1084,14 +1084,14 @@ public ManifestsResponseDTO getManifestIdentifiers(String yaml, String serviceId
             String.format("Yaml provided for service %s does not have service root field.", serviceIdentifier));
       }
```
Comment (tchaurasiya) [inline 127-cd-nextgen-entities/src/main/java/io/harness/ng/core/service/services/impl/ServiceEntityServiceImpl.java:1085]: discuss internally if this should be added to spec first api not current service resourceV2 api. Also lets decide on if we should throw exception for not finding manifest in service

**Review by tchaurasiya — CHANGES_REQUESTED** [review]

**Comment by GoelPratyush99** [comment]
trigger uts

**Comment by GoelPratyush99** [comment]
trigger uts

**Comment by GoelPratyush99** [comment]
trigger uts

**Review by yogesh-chauhan — APPROVED** [review]

**Comment by GoelPratyush99** [comment]
trigger uts

**Review by yogesh-chauhan — COMMENTED** [review]
> Reviewed 1 of 1 files at r6, 1 of 4 files at r7, 10 of 12 files at r9, 2 of 3 files at r10, 1 of 1 files at r11, 4 of 7 files at r12, 3 of 3 files at r14, 1 of 1 files at r15, 1 of 1 files at r16, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/50559)* status: 10 file(s) to review. 2 change requests. 5. unresolved discussions. 0 of 2 approvals obtained.

**Review by yogesh-chauhan — APPROVED** [review]

**Review by abosii-harness — APPROVED** [review]

**Review by tchaurasiya — APPROVED** [review]

**Review by TarunUba1 — APPROVED** [review]

**Comment by GoelPratyush99** [comment]
trigger review
